Add the `size` API for TensorFlow NumPy operations.

DarrenZhang01 opened the pull request:
The equivalent vanilla NumPy operation: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.size.html

Commit:
Add the `size` API for TensorFlow NumPy operations.

Commit:
Remove the `print` statement.

Commit:
Revise the code format according to pylint.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -560,6 +560,19 @@ def _reduce(tf_fn,
 
+@utils.np_doc('size')
+def size(x):
+  x = x.numpy() if isinstance(x, tf.Tensor) else x
```
This currently won't handle graph tensors. How about something similar to below to handle those as well?

```suggestion
  x = asarray(x).data
  sz_tensor = array_ops.size(x)
  sz_value = np_utils.get_static_value(sz_tensor)
  if sz_value is not None:
    return int(sz_value)
  return np_utils.tensor_to_ndarray(sz_tensor)
```

DarrenZhang01 replied:
Thanks for the insight! Just to make sure, can we turn a graph Tensor directly into TF NumPy array?

akshaym replied:
Yes! The ndarray class can wrap either graph or eager tensors.

DarrenZhang01 replied:
Thanks Akshay! I will commit this a bit later since I am currently dealing with some Neural Tangents issues.

akshaym replied:
Unresolving this since it looks like the code still doesn't handle graph tensors.

DarrenZhang01 replied:
Hi Akshay @akshaym , thanks for pointing this out! I further revised the code in https://github.com/tensorflow/tensorflow/pull/42539/commits/fc5de5045086933fa7a657f0133b353a80b5d04a and now it should work fine.

akshaym replied:
Hi Zhibo, the code as written still doesn't handle graph tensors. 

Graph tensors can have undefined shapes during graph construction (so for example the tensor can have a shape of "[None, None]")

In this situation [tnp.ndarray.shape](https://github.com/tensorflow/tensorflow/blob/a067c6723019ea08d3b4a1acf2cb1dbd42e33e09/tensorflow/python/ops/numpy_ops/np_arrays.py#L197) will return a tf.Tensor, which you cannot pass to np.prod.

DarrenZhang01 replied:
Thanks for mentioning this, Akshay. In your opinion, how should we handle undefined shapes ... raise `ValueError`?

DarrenZhang01 replied:
By the way, would you mind providing a concrete example on what a `None` shape Tensor construction is like? Thanks.

akshaym replied:
The suggestion I gave above (in the first comment on this thread) should handle both graph and eager tensors. 

Here is a contrived example:
```
@tf.function(input_signature=[tf.TensorSpec(shape=None, dtype=tf.float32)])
def f(x):
  x = tnp.asarray(x)  # Wraps a graph tensor into a tnp.ndarray
  return tnp.size(x)  # This should fail since the shape is "None"
```

DarrenZhang01 replied:
Hi Akshay, thanks for the insight! I tried wrapping the `tf.TensorSpec(shape=None, dtype=tf.float32)` using `np.asarray`, but got the error `TypeError: data type not understood`.

DarrenZhang01 replied:
Hi @akshaym , I revised the code in https://github.com/tensorflow/tensorflow/pull/42539/commits/d70b16b8a54d15afc9b0fc33f0bbe93ba22ba456 based on your suggestion.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops_test.py:
```diff
@@ -804,6 +804,22 @@ def testSum(self):
+
+    def run_test(arr):
+      onp_arr = arr.numpy() if isinstance(arr, tf.Tensor) else arr
+      self.assertEqual(np_size(arr), onp.size(onp_arr))
```
Should this be np_array_ops.size(arr)?

DarrenZhang01 replied:
Yeah, sorry, `np_size` is my local version when testing.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -560,6 +560,19 @@ def _reduce(tf_fn,
 
 
+@utils.np_doc('size')
+def size(x):
```
Looking at the documentation for np.size, it seems to have an axis argument as well. Can you add that as well? (We don't have to support the code for now, we can simply raise an error if the 'axis' argument is something other None)

Please add a test for that as well.

DarrenZhang01 replied:
Hi, Akshay! Should I also include tests for raising errors?

DarrenZhang01 replied:
Never mind. I included the according tests.

akshaym reviewed (commented):
Thanks for the changes @DarrenZhang01!

Commit:
1. Use `np.asarray` wrapper to process the input in order to be able to
     deal with grapg tensors;
  2. Add the `axis` parameter into the `size` API, raise
     `NotImplementedError` when the not-None `axis` parameter is sent
     in. Add the according tests.

Commit:
1. Add `# pylint: disable=missing-docstring` label beside the function;
2. Revise the name of TF NumPy and vanilla NumPy according to the
   imports.

Commit:
Revise `utils` to `np_utils`.

Commit:
Revise the bug in the test code.

Commit:
Improve the `size` API and revise some bugs.

Commit:
Revise the code to handle the graph tensors.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -560,6 +561,23 @@ def _reduce(tf_fn,
+  if axis is not None:
+    raise NotImplementedError("axis argument is not supported in the current "
+                              "`np.size` implementation")
+  if isinstance(x, (int, float, np.int32, np.int64,
```
After your change above, this should not happen. `x` will **always** be a Tensor. You can move this above the `asarray` call, since I think it makes sense as an optimization.

DarrenZhang01 replied:
Great idea!

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -560,6 +561,23 @@ def _reduce(tf_fn,
+                    np.float32, np.float64)):
+    return 1
+  elif isinstance(x, (np_arrays.ndarray, np.ndarray)) or tf.is_tensor(x):
+    return prod(tuple(x.shape))
```
This still doesn't handle Tensors with partially defined shapes. 

x.shape returns a [TensorShape](https://www.tensorflow.org/api_docs/python/tf/TensorShape). If the shape [is_fully_defined](https://www.tensorflow.org/api_docs/python/tf/TensorShape#is_fully_defined), then the code as written works OK, but that is not guaranteed. If the shape isn't fully defined, you'll have to call the corresponding TensorFlow [size](https://www.tensorflow.org/api_docs/python/tf/size) function, and return that as a result.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -24,6 +24,7 @@
 import numpy as np
 import six
 
+import tensorflow as tf
```
We never `import tensorflow as tf` from within the TensorFlow codebase. We should be importing the files that define the functions we want (looks like you want `from tensorflow.python.framework import tensor_util`), and use them directly like other imports in this file (specifically here `tensor_util.is_tensor`). 

You'll probably want to remove it from the test file as well.

Commit:
1. Use tf.size to deal with None shapes;
2. Change from external import to intern import.

DarrenZhang01 commented:
Hi Akshay @akshaym , I have revised the places as you suggested. I also tried to add tests for the `None` shape case. However, when I tried to define a `TensorSpec` as you suggested and converted using `asarray` wrapper, it always gives me the error `TypeError: data type not understood`.

Commit:
Import `sys` at the original place in order to avoid unnecessary
changes.

akshaym commented:
Hi @DarrenZhang01, `asarray` cannot wrap a TensorSpec.

If you want to test the code for partially defined shapes, you can try something like the following (copying the code I shared previously):
```
@tf.function(input_signature=[tf.TensorSpec(shape=None, dtype=tf.float32)])
def f(x):
  x = tnp.asarray(x)  # Wraps a graph tensor into a tnp.ndarray
  return tnp.size(x)  # This should fail since the shape is "None"
```
Note that this code doesn't ever run `asarray` on a TensorSpec, it runs `asarray` on the input `x` (which is a graph tensor that has the shape `None`, and dtype `float32`).

You can now call the function `f` defined above with e.g. `f(tnp.ones([2, 3])`. If you try the above code exactly without your latest changes for supporting partially defined shapes, it should fail.

Commit:
Add test case for the `None` shape case as pointed out in
https://github.com/tensorflow/tensorflow/pull/42539#issuecomment-686715283.

DarrenZhang01 commented:
> Hi @DarrenZhang01, `asarray` cannot wrap a TensorSpec.
> 
> If you want to test the code for partially defined shapes, you can try something like the following (copying the code I shared previously):
> 
> ```
> @tf.function(input_signature=[tf.TensorSpec(shape=None, dtype=tf.float32)])
> def f(x):
>   x = tnp.asarray(x)  # Wraps a graph tensor into a tnp.ndarray
>   return tnp.size(x)  # This should fail since the shape is "None"
> ```
> 
> Note that this code doesn't ever run `asarray` on a TensorSpec, it runs `asarray` on the input `x` (which is a graph tensor that has the shape `None`, and dtype `float32`).
> 
> You can now call the function `f` defined above with e.g. `f(tnp.ones([2, 3])`. If you try the above code exactly without your latest changes for supporting partially defined shapes, it should fail.

Thanks very much, Akshay @akshaym ! I got what you said. Now the code is again ready for review.

akshaym reviewed (commented):
Thanks for the changes @DarrenZhang01! Just some small nits left now.

Commit:
Remove some unnecessary `if...else...` statement; Wrap the return value
with an ndarray converter.

DarrenZhang01 commented:
Thanks very much, @akshaym ! I learned a lot!

akshaym reviewed (dismissed):
Thanks for sticking through all the back and forth @DarrenZhang01!

akshaym commented:
Hey Zhibo @DarrenZhang01, Since you're adding a new API endpoint, you'll also need to regenerate the API goldens. 

You'll have to run 
`bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True`

DarrenZhang01 commented:
> Hey Zhibo @DarrenZhang01, Since you're adding a new API endpoint, you'll also need to regenerate the API goldens.
> 
> You'll have to run
> `bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True`

Akshay, where do I run this command?

akshaym commented:
In the top-level directory of your tensorflow repo.

DarrenZhang01 commented:
> In the top-level directory of your tensorflow repo.

@akshaym A small question: Do I have to downgrade bazel to 3.1.0 (I have not found a proper way to do so on Mac)? If I build TensorFlow using bazel 3.5 (latest), the running process seems to be endless...

DarrenZhang01 commented:
Hi Akshay @akshaym , as a follow-up to the above message, even if I use bazel 3.1.0, the build process seems to be endless:

```
INFO: Analyzed target //tensorflow/tools/api/tests:api_compatibility_test (366 packages loaded, 31781 targets configured).
INFO: Found 1 target...
[7,522 / 8,644] 4 actions running
    Compiling tensorflow/core/kernels/relu_op.cc; 48s local
    Compiling tensorflow/core/kernels/linalg/matrix_inverse_op.cc; 26s local
    Compiling tensorflow/core/kernels/gather_nd_op.cc; 14s local
    Compiling tensorflow/core/kernels/gather_nd_op_cpu_impl_0.cc; 7s local
```
i.e., the amount number right to the slash, 8644, is increasing all the time...

Commit:
Add the change after running bazel.

akshaym commented on tensorflow/python/ops/numpy_ops/np_array_ops.py:
```diff
@@ -28,6 +28,7 @@
 from tensorflow.python.framework import dtypes
 from tensorflow.python.framework import ops
 from tensorflow.python.framework import tensor_shape
+from tensorflow.python.framework import tensor_util
```
Hey @DarrenZhang01 I think this needs to be removed since its no longer used.

DarrenZhang01 replied:
Thanks for pointing this out, Akshay!

Commit:
Remove the unnecessary import based on
`https://github.com/tensorflow/tensorflow/pull/42539#discussion_r485189850`.
